### PR TITLE
use hyperium/http instead of http-rs/http-types

### LINKED ISSUE #123: http-types dependency seems unmaintained
Hi! Is it important for wiremock to use the `http-types` crate (from [http-rs](https://github.com/http-rs) which unfortunately looks pretty dead as a whole) as opposed to the `http` crate (from hyperium)? It has not seen a release in almost two years and pulls in several unwanted dependencies ([old] `serde_qs`, old `rand`).

### PR DESCRIPTION
**Pull request by flisky**
Fix #123 

Please attention: the exported types `Method` / `HeaderValues` make this a breaking change.

By the way, `hyper` v1.0 will remove `hyper::Body`, but I think it's fine to keep it for now?

--- 
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Cargo.toml
Comment (LukeMathWalker) [inline Cargo.toml:43]: We want to keep testing that we work on all these runtimes.

Comment (LukeMathWalker) [reply Cargo.toml:43]: The test suite shouldn't be touched beyond the changes forced by API differences.

Comment (flisky) [reply Cargo.toml:43]: a16eb6ee05fadc218f36c2a9821458a1c5ae2600 was introduced mainly because I suppose `http-rs/surf` is not active mantained as `http-rs/http-types`?

Re: we work on all these runtimes/HTTP clients
I don't think it's a problem, because we use `hyper` as a mock server and `http` as transport protocol, which is quite standard and all HTTP clients should obey.

It's better to review commit by commit. The first commit did most of the work except for doc tests, and the remaining commits are housekeeping.

Comment (LukeMathWalker) [reply Cargo.toml:43]: There are sharp edges in how things interoperate across runtimes, which is why those tests were there in the first place using those specific libs. There is no problem in keeping a not-very-actively-maintained library as a dev dependency.

**Comment by flisky** [comment]
friendly ping

**Comment by LukeMathWalker** [comment]
> friendly ping

Friendly tip: never, ever, do this on an OSS project tracker 😄

I review PRs when I have the capacity to, pinging me doesn't make me review it any sooner.

**Comment by jplatte** [comment]
> > friendly ping
> 
> Friendly tip: never, ever, do this on an OSS project tracker 😄

While I agree that the above ping 2d after the last update was not very helpful, as a maintainer of many projects I want to note that forgetting about a PR is a thing I've been glad that people pinged me after a few weeks of inactivity on multiple occasions.

File: src/matchers.rs
Comment (LukeMathWalker) [inline src/matchers.rs:398]: We can't unwrap here.

Comment (LukeMathWalker) [reply src/matchers.rs:398]: It's fine to unwrap in matcher creation, since that would bubble up into the test setup code. But we can't unwrap in the matching itself, otherwise that's going to be pretty confusing to debug.

File: src/matchers.rs
Comment (LukeMathWalker) [inline src/matchers.rs:403]: This has a different semantic compared to the previous one: this is order-insensitive, while the previous implementation was order-sensitive.

File: src/mock.rs
Comment (LukeMathWalker) [inline src/mock.rs:24]: We could go for `.len()` directly on the header value here.

**Comment by LukeMathWalker** [comment]
> > > friendly ping
> > 
> > 
> > Friendly tip: never, ever, do this on an OSS project tracker 😄
> 
> While I agree that the above ping 2d after the last update was not very helpful, as a maintainer of many projects I want to note that forgetting about a PR is a thing I've been glad that people pinged me after a few weeks of inactivity on multiple occasions.

A few weeks is fine, but 2 days is not—I don't even ping co-workers that soon 😅

File: src/mock_server/bare_server.rs
Comment (flisky) [reply src/mock_server/bare_server.rs:83]: see 98a69d9f0ea849e17c8a0877b61d985b223f2306, for less dependency

File: src/request.rs
Comment (jplatte) [inline src/request.rs:57]: Just my 2c, I'd do this as two statements.

File: src/response_template.rs
Comment (jplatte) [inline src/response_template.rs:260]: The `Response::builder` API should make this code a bit shorter.

**Review by jplatte — APPROVED** [review]
Not sure how useful an external review is, but apart from the header matching that I haven't dug into, this looks good to me. Only have two very minor comments.

File: src/respond.rs
Comment (LukeMathWalker) [inline src/respond.rs:78]: Can't we import from `http` here? I don't think we need to import from `hyper`.

**Comment by LukeMathWalker** [comment]
This looks good to go, just some conflicts to resolve—thanks for the work!
I want to wait for `hyper` to push `1.0` out before merging and releasing this though.

File: src/respond.rs
Comment (jplatte) [reply src/respond.rs:78]: I think it would require an extra (dev) dependency, but probably worth it.

Comment (LukeMathWalker) [reply src/respond.rs:78]: It would, but `http` is already in our (non-dev) dependency tree, so it doesn't make an actual different.

Comment (jplatte) [reply src/respond.rs:78]: Yeah, just one more version to keep in sync with the test in the manifest.

**Comment by LukeMathWalker** [comment]
Both `hyper` and `http` cut 1.0—let's update and then we can merge and release!

**Comment by flisky** [comment]
Glad to hear! Will give it a try on this weekends.

**Comment by flisky** [comment]
So here is the hyper v1 upgrade.

However, I have to use a **multi-threaded** runtime to run the tests (see my last commit). 
The `spawn_local` is just not scheduled at all even for a single testcase.(It's okay if I use `tokio::net::TcpListener` from top to down.)
I take hours to investigate, but no clue. Could you give some sight, please? @LukeMathWalker @jplatte Thanks!

**Comment by LukeMathWalker** [comment]
I discovered a while ago that converting a `std` `TcpListener` into a `tokio` `TcpListener` won't set it to non-blocking. I wonder if that's the issue here?

**Comment by flisky** [comment]
Yes, it's exactly the problem! I push another commit to switch back to single-thread runtime, thanks!

File: src/http.rs
Comment (jplatte) [inline src/http.rs:3]:
```suggestion
pub use http::{HeaderMap, HeaderName, HeaderValue, Method};
```

File: src/matchers.rs
Comment (jplatte) [inline src/matchers.rs:14]:
```suggestion
use http::{HeaderName, HeaderValue, Method};
```

File: src/request.rs
Comment (jplatte) [inline src/request.rs:4]:
```suggestion
use http::{HeaderMap, Method};
```

File: src/response_template.rs
Comment (jplatte) [inline src/response_template.rs:4]:
```suggestion
use http::{HeaderMap, HeaderName, HeaderValue, Response, StatusCode};
use http_body_util::Full;
use hyper::body::Bytes;
```

**Review by jplatte — APPROVED** [review]

**Pull request merged**
